File: app/src/main/java/com/example/changfeng/taptapword/SettingsFragment.java
package com.example.changfeng.taptapword;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Environment;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;


import java.io.File;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;


public class SettingsFragment extends Fragment implements View.OnClickListener {

    private static final String TAG = "SettingsFragment";
    static final int REQUEST_BACKUP_FILE = 1;
    private static final String FILE_KEY_WORD = "word_ninja";



    @Bind(R.id.youdao_dict_switch)
    Switch youdaoDictSwitch;
    @Bind(R.id.web_explain_switch)
    Switch webExplainSwitch;
    @Bind(R.id.youdao_translate_switch)
    Switch youdaoTranslateSwitch;
    @Bind(R.id.baidu_translate_switch)
    Switch baiduTranslateSwitch;

    @Bind(R.id.backup_data)
    TextView backupDataTextView;
    @Bind(R.id.restore_data)
    TextView restoreDataTextView;
    @Bind(R.id.clear_backup_data)
    TextView clearBackupDataTextView;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_settings, container, false);
        ButterKnife.bind(this, view);

        backupDataTextView.setOnClickListener(this);
        restoreDataTextView.setOnClickListener(this);
        clearBackupDataTextView.setOnClickListener(this);



        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        readSettings();
    }

    @Override
    public void onPause() {
        super.onPause();
        saveSettings();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {

            case R.id.backup_data:
                new AlertDialog.Builder(getActivity())
                        .setTitle(R.string.backup)
                        .setMessage(R.string.message_backup_data)
                        .setPositiveButton(R.string.button_ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                if (WordManger.get(getActivity()).copyDbToSdcard()) {
                                    showToast(R.string.message_data_backupped);
                                } else {
                                    showToast(R.string.message_data_backupped_failed);
                                }
                            }
                        })
                        .setCancelable(true)
                        .create().show();
                break;
            case R.id.restore_data:
                new AlertDialog.Builder(getActivity())
                        .setTitle(R.string.restore)
                        .setMessage(R.string.message_restore_data)
                        .setPositiveButton(R.string.button_ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                Intent intent = new Intent(getActivity(), FileListActivity.class);
                                intent.putExtra(FileListActivity.FILE_KEY_WORD, "word_ninja");
                                startActivityForResult(intent, REQUEST_BACKUP_FILE);
                            }
                        })
                        .setCancelable(true)
                        .create().show();
                break;
            case R.id.clear_backup_data:
                new AlertDialog.Builder(getActivity())
                        .setTitle(R.string.clear_backup_data)
                        .setMessage(R.string.message_clear_backup_data)
                        .setPositiveButton(R.string.button_ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                clearAllBackupData();
                                showToast(getString(R.string.message_all_backup_data_cleared), Toast.LENGTH_SHORT);
                            }
                        })
                        .setCancelable(true)
                        .create().show();
                break;
            default:
                break;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
//        Log.d(TAG, "onActivityResult() called resultCode :" + requestCode + " requestCode :" + requestCode);
        if (resultCode == Activity.RESULT_OK) {
            switch (requestCode) {
                case REQUEST_BACKUP_FILE:
                    if (WordManger.get(getActivity()).restoreDb(data.getExtras().getString(FileListActivity.FILE_KEY_WORD))) {
                        showToast(getString(R.string.message_data_restored));
                    } else {
                        showToast(getString(R.string.message_data_restored_failed));
                    }
                    break;
                default:
                    break;
            }
        }
    }

    private void clearAllBackupData() {
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            File path = Environment.getExternalStorageDirectory();
            File files[] = path.listFiles();
            if (files != null) {
                for (File file : files) {
                    if (file.isDirectory()) {
                        continue;
                    }
                    if (file.getName().contains(FILE_KEY_WORD)) {
                        MyFile.deleleFile(file.getAbsolutePath());
                    }
                }
            }
        }
    }

    private void readSettings(){
        SharedPreferences pref  = getActivity().getSharedPreferences(SharedPref.PREFERENCE_NAME, Context.MODE_PRIVATE);
        youdaoDictSwitch.setChecked(pref.getBoolean(SharedPref.PREFERENCE_YOUDAO_DICT, true));
        webExplainSwitch.setChecked(pref.getBoolean(SharedPref.PREFERENCE_WEB_EXPLAIN, true));
        youdaoTranslateSwitch.setChecked(pref.getBoolean(SharedPref.PREFERENCE_YOUDAO_TRANSLATE, true));
        baiduTranslateSwitch.setChecked(pref.getBoolean(SharedPref.PREFERENCE_DAIDU_TRANSLATE, true));
    }

    private void saveSettings() {
        SharedPreferences.Editor editor = getActivity().getSharedPreferences(SharedPref.PREFERENCE_NAME, Context.MODE_PRIVATE).edit();
        editor.putBoolean(SharedPref.PREFERENCE_YOUDAO_DICT, youdaoDictSwitch.isChecked())
                .putBoolean(SharedPref.PREFERENCE_WEB_EXPLAIN, webExplainSwitch.isChecked())
                .putBoolean(SharedPref.PREFERENCE_YOUDAO_TRANSLATE, youdaoTranslateSwitch.isChecked())
                .putBoolean(SharedPref.PREFERENCE_DAIDU_TRANSLATE, baiduTranslateSwitch.isChecked())
                .apply();
    }



    private void showToast(int resourceId) {
        Toast.makeText(getActivity(), getString(resourceId), Toast.LENGTH_SHORT).show();
    }

    private void showToast(String message) {
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
    }

    private void showToast(String message, int duration) {
        Toast.makeText(getActivity(), message, duration).show();
    }
}

File: app/src/main/java/com/example/changfeng/taptapword/SettingsFragment.kt
package com.example.changfeng.taptapword


import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Switch
import android.widget.TextView
import androidx.fragment.app.Fragment
import org.jetbrains.anko.*
import org.jetbrains.anko.support.v4.act
import org.jetbrains.anko.support.v4.find
import org.jetbrains.anko.support.v4.toast


class SettingsFragment : Fragment(), View.OnClickListener {
    val youdaoDictSwitch: Switch get() = find(R.id.youdao_dict_switch)
    val webExplainSwitch: Switch get() = find(R.id.web_explain_switch)
    val youdaoTranslateSwitch: Switch get() = find(R.id.youdao_translate_switch)
    val baiduTranslateSwitch: Switch get() = find(R.id.baidu_translate_switch)
    val backupDataTextView: TextView get() = find(R.id.backup_data)
    val restoreDataTextView: TextView get() = find(R.id.restore_data)
    val clearBackupDataTextView: TextView get() = find(R.id.clear_backup_data)
    val ninjaWatchNotificationSwitch: Switch get() = find(R.id.notification_ninja_watch)
    val newWordNotificationSwitch: Switch get() = find(R.id.notification_new_word)


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_settings, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        backupDataTextView.setOnClickListener(this)
        restoreDataTextView.setOnClickListener(this)
        clearBackupDataTextView.setOnClickListener(this)

    }

    override fun onResume() {
        super.onResume()
        readSettings()
    }

    override fun onPause() {
        super.onPause()
        saveSettings()
    }

    override fun onClick(v: View) {
        when (v.id) {

            R.id.backup_data -> {
                requireActivity().alert(R.string.message_backup_data, R.string.backup) {
                    yesButton {
                        if (WordManger.get(act).copyDbToSdcard()) {
                            toast(R.string.message_data_backupped)
                        } else {
                            toast(R.string.message_data_backupped_failed)
                        }
                    }
                    noButton { }
                }.show()
            }

            R.id.restore_data -> {
                requireActivity().alert(R.string.message_restore_data, R.string.restore) {
                    yesButton {
                        startActivityForResult(requireActivity().intentFor<FileListActivity>(FILE_KEY_WORD to "word_ninja"), REQUEST_BACKUP_FILE)
                    }
                    noButton { }
                }.show()
            }
            R.id.clear_backup_data ->
                requireActivity().alert(R.string.message_clear_backup_data, R.string.clear_backup_data) {
                    yesButton {
                        clearAllBackupData()
                        toast(R.string.message_all_backup_data_cleared)
                    }
                    noButton {  }
                }.show()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != Activity.RESULT_OK) {
            return
        }

        when (requestCode) {
            REQUEST_BACKUP_FILE -> if (WordManger.get(activity).restoreDb(data!!.extras!!.getString(FileListActivity.FILE_KEY_WORD))) {
                toast(R.string.message_data_restored)
            } else {
                toast(R.string.message_data_restored_failed)
            }
            else -> {
            }
        }
    }

    private fun clearAllBackupData() {
        if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
            val path = Environment.getExternalStorageDirectory()
            path.listFiles { _, s -> s.contains(FILE_KEY_WORD) }.filter { it -> it.isFile }.forEach { it.delete() }
        }
    }

    private fun readSettings() {
        val pref = requireActivity().defaultSharedPreferences
        youdaoDictSwitch.isChecked = pref.getBoolean(SharedPref.YOUDAO_DICT, true)
        webExplainSwitch.isChecked = pref.getBoolean(SharedPref.WEB_EXPLAIN, true)
        youdaoTranslateSwitch.isChecked = pref.getBoolean(SharedPref.YOUDAO_TRANSLATE, true)
        baiduTranslateSwitch.isChecked = pref.getBoolean(SharedPref.BAIDU_TRANSLATE, true)
        ninjaWatchNotificationSwitch.isChecked = pref.getBoolean(SharedPref.NOTIFICATION_NINJA_WATCH, true)
        newWordNotificationSwitch.isChecked = pref.getBoolean(SharedPref.NOTIFICATION_NEW_WORD, true)
    }

    private fun saveSettings() {
        val editor = requireActivity().defaultSharedPreferences.edit()
        editor
            .putBoolean(SharedPref.YOUDAO_DICT, youdaoDictSwitch.isChecked)
            .putBoolean(SharedPref.WEB_EXPLAIN, webExplainSwitch.isChecked)
            .putBoolean(SharedPref.YOUDAO_TRANSLATE, youdaoTranslateSwitch.isChecked)
            .putBoolean(SharedPref.BAIDU_TRANSLATE, baiduTranslateSwitch.isChecked)
            .putBoolean(SharedPref.NOTIFICATION_NINJA_WATCH, ninjaWatchNotificationSwitch.isChecked)
            .putBoolean(SharedPref.NOTIFICATION_NEW_WORD, newWordNotificationSwitch.isChecked)
            .apply()
    }

    companion object {

        private val TAG = "SettingsFragment"
        internal val REQUEST_BACKUP_FILE = 1
        private val FILE_KEY_WORD = "word_ninja"
    }
}